enhancements/cluster-runtime-constraints.md

```diff
@@ -0,0 +1,138 @@
+
+### User Stories
+
+#### Constraints specified by cluster admin
+
+As a cluster admin, I would like to specify a list of cluster runtime constraints that are associated with a given cluster and all installable operators will meet these requirements.
+
+#### Constraint properties specified by operator author
+
+As an operator author, I would like to provide a list of runtime properties that may be needed to be considered before installation to ensure the operator working properly on the cluster.
+
+### Implementation Details
+
+#### Cluster Runtime Constraint Configmap
+
+Cluster admins can specified cluster runtime constraints in a configmap (named `olm-runtime-constraints`) in `olm` namespace using this format:
+
+```yaml=
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: olm-runtime-constraints
+  namespace: olm
+data:
+    properties: <list-of-expression-constraints-in-json>
+```
+
+Each constraint is specified using the following syntax:
+
+```yaml=
+{
+  "type": "olm.constraint",
+  "value": {
+    "evaluator": {
+        "id": "cel",
+    }
+    "source": 'properties.exists(p, p.type == "certified")',
+    "action": {
+        "id": "require"
+    }
+  }
+}
+```
+
+The constraint `type` is `olm.constraint` and the `value` is information associated with constraint. The `value` struct has 3 fields: `evaluator`, `source` and `action`.
+
+The `evaluator` is a struct with `id` field to represent the languague library that will be used to evaluate the expression. At the moment, only `cel-go` library is supported using `cel` as the identifier. More fields are potentially added to expand the supported expression languagues/libraries in the future.
+
+The `source` field is the string of expression that will be evaluated during resolution. Currently, only `cel-go` expressions are supported.
+
+Finally, the `action` field is a struct with the `id` field to indicate the resolution action that this constraint will behave. For example, for `require` action, there must be at least one candidate that satisfies this constraint. At the moment, `require` and `conflict` actions are supported. More fields may be added into `action` struct to support more resolution actions in the future.
+
+#### Runtime Dependencies in Bundle
+
+The operator authors specify the cluster runtime requirements in `dependencies.yaml` using the following format:
+
+```yaml=
+dependencies:
+  -
+    type: olm.constraint
+    value:
+      action:
+        id: require
+      evaluator:
+        id: cel
+      source: "properties.exists(p, p.type == \"certified\")"
+```
+
+#### Runtime Constraint Resolution
+
+The constraints that are specified in `olm-runtime-constraints` Configmap are converted into properties of a global existing virtual node in the cluster. All operators/candidates must satisfy those properties in order to be installed.
```

Comment:
What exactly is "a global existing virtual node in the cluster" ?

Reply:
It essentially means an existing global operator.

Reply:
Are there any rough edges around making the cluster provided properties just another node in the SAT solver?

What if an arbitrary bundle includes a "kubernetes version" property? Is there any notion of property priority or mutual exclusivity for property providers?

Can cluster admins see what properties exist on a bundle when they are added? If not, how will we help cluster admins prevent installation of "malicious" bundles that populate unwanted properties in a cluster?